### ISSUE TITLE
Random segfaults with e2guardian 4.1.1 

### ISSUE THREAD

**Issue by fredbcode** [e2guardian/e2guardian#253]
#0  __memcpy_sse2_unaligned () at ../sysdeps/x86_64/multiarch/memcpy-sse2-unaligned.S:36
#1  0x00007f8cbff42650 in char* std::string::_S_construct<char*>(char*, char*, std::allocator<char> const&, std::forward_iterator_tag) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#2  0x00007f8cbff426f1 in std::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(std::string const&, unsigned long, unsigned long) ()
   from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007f8cbff42735 in std::string::substr(unsigned long, unsigned long) const () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x0000000000409218 in String::after(char const*) const ()
#5  0x000000000042c735 in HTTPHeader::contentLength() ()
#6  0x0000000000417ddd in ConnectionHandler::handleConnection(Socket&, String&, bool, Socket&, stat_rec*&) ()
#7  0x000000000040c6ef in ConnectionHandler::handlePeer(Socket&, String&, stat_rec*&) ()
#8  0x000000000044b9c8 in handle_connections(int) ()
#9  0x000000000045ba04 in void std::_Bind_simple<void (*(int))(int)>::_M_invoke<0ul>(std::_Index_tuple<0ul>) ()
#10 0x000000000045b883 in std::_Bind_simple<void (*(int))(int)>::operator()() ()
#11 0x000000000045b77e in std::thread::_Impl<std::_Bind_simple<void (*(int))(int)> >::_M_run() ()
#12 0x00007f8cbff38970 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#13 0x00007f8cc0a0d064 in start_thread (arg=0x7f8c9dffb700) at pthread_create.c:309
#14 0x00007f8cbf6a862d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111

In guess the problem is in String.cpp with some not printable characters 

```
String::after(const char *bs) const

String String::after(const char *bs) const
{
    size_type i = this->find(bs);
    if (i == std::string::npos)
        return "";
    return this->substr(i + strlen(bs));
}
```

substr seems produce a segfault

A way to reproduce 
With a high load pkill -9 squid and wait some minutes after a while e2guardian (sometimes ...) crash
With some debug values I saw a no printable characters (like binary ?) to the string, but I'm not 100 % sure this is related, the core generated is always the same

**Comment by fredbcode** [comment]
Same crash, without squid restart

```
#0  __memcpy_sse2_unaligned () at ../sysdeps/x86_64/multiarch/memcpy-sse2-unaligned.S:36
#1  0x00007f452d4e0650 in char* std::string::_S_construct<char*>(char*, char*, std::allocator<char> const&, std::forward_iterator_tag) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#2  0x00007f452d4e06f1 in std::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(std::string const&, unsigned long, unsigned long) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007f452d4e0735 in std::string::substr(unsigned long, unsigned long) const () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x0000000000409218 in String::after(char const*) const ()
#5  0x000000000042c658 in HTTPHeader::returnCode() ()
#6  0x000000000041240a in ConnectionHandler::handleConnection(Socket&, String&, bool, Socket&, stat_rec*&) ()
#7  0x000000000040c6ef in ConnectionHandler::handlePeer(Socket&, String&, stat_rec*&) ()
#8  0x000000000044b9c8 in handle_connections(int) ()
#9  0x000000000045ba04 in void std::_Bind_simple<void (*(int))(int)>::_M_invoke<0ul>(std::_Index_tuple<0ul>) ()
#10 0x000000000045b883 in std::_Bind_simple<void (*(int))(int)>::operator()() ()
#11 0x000000000045b77e in std::thread::_Impl<std::_Bind_simple<void (*(int))(int)> >::_M_run() ()
#12 0x00007f452d4d6970 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#13 0x00007f452dfab064 in start_thread (arg=0x7f3314182700) at pthread_create.c:309
#14 0x00007f452cc4662d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

**Comment by marcelloc** [comment]
Check if they are related to ssl communications. Maybe not the only crash needs attention but on my production boxes, I need to enable splice all on parent cache to keep e2guardian running. 
If I disable splice all, this high load server crashes every 30seconds depending on traffic.

**Comment by fredbcode** [comment]
I'm running without SSLMITM (Squid and E2guardian) and I have one or two crashes each day, often very close after a squid reload/restart.

I have more or least 2000 simultaneous users connected (httpsworkers = 8000)

> this high load server crashes every 30seconds depending on traffic. 

Can you setup your environment to create a core dump ? (there are some words for Debian on wiki) like this we can see if our issues are related or not.

As far I can tell there is no (yet ?) report for such massive problem

>Check if they are related to ssl communications 

Yes I will try but hard to say now, too much simultaneous connections
But there is a "good" news , for the moment the core result is always about the same thing

```
#3  0x00007f452d4e0735 in std::string::substr(unsigned long, unsigned long) const () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x0000000000409218 in String::after(char const*) const ()
#5  0x000000000042c658 in HTTPHeader::returnCode() () 
```

**Comment by fredbcode** [comment]
I forgot Squid 3.5.25

**Comment by marcelloc** [comment]
I'm on 4.1.2-develop branch

**Comment by fredbcode** [comment]
Yes me top dev branch

**Comment by fredbcode** [comment]
I missed a fix from Philip, can make a try with the latest push in develop

**Comment by fredbcode** [comment]
Fixed in 4.1.2
